store/store.go

```diff
@@ -500,3 +500,50 @@ func mustEncode(pb proto.Message) []byte {
 	}
 	return bz
 }
+
+//-----------------------------------------------------------------------------
+
+// DeleteLatestBlock removes the block pointed to by height,
+// lowering height by one.
+func (bs *BlockStore) DeleteLatestBlock() error {
+	bs.mtx.RLock()
+	targetHeight := bs.height
+	bs.mtx.RUnlock()
+
+	batch := bs.db.NewBatch()
+	defer batch.Close()
+
+	// delete what we can, skipping what's already missing, to ensure partial
+	// blocks get deleted fully.
+	if meta := bs.LoadBlockMeta(targetHeight); meta != nil {
+		if err := batch.Delete(calcBlockHashKey(meta.BlockID.Hash)); err != nil {
+			return err
+		}
+		for p := 0; p < int(meta.BlockID.PartSetHeader.Total); p++ {
+			if err := batch.Delete(calcBlockPartKey(targetHeight, p)); err != nil {
+				return err
+			}
+		}
+	}
+	if err := batch.Delete(calcBlockCommitKey(targetHeight)); err != nil {
+		return err
+	}
+	if err := batch.Delete(calcSeenCommitKey(targetHeight)); err != nil {
+		return err
+	}
```

Comment:
Is it dangerous to do delete this? We are losing the seen commit. Consensus code relies on this being set so it seems risky.

Reply:
It's only the seen commit for that height. Remember this is before we decided to use only a single seen commit. The node will restart and have to sync up. By then it will have the seen commit at the lastest height it needs in the case it's the proposer in the following height

Reply:
This will also have implications in the optimal solution for vote extension propagation described as "Solution 3" in RFC 017. But we're a long way from there (we'd first need to be able to switch from consensus back to blocksync).
So take this just as a heads-up :-)